### ISSUE TITLE
Datetime fields given relative formats without a reference point always use current system time - should use Datetime::now() as reference point?

### ISSUE THREAD

**Issue by p0lar-bear** [silverstripe/silverstripe-framework#8010]
## Affected Version

3.5, likely 3.6, possibly 4.x
PHP 5.6

## Description

When an `SS_Datetime` field is given a relative format as outlined in [PHP: Relative Formats](http://php.net/manual/en/datetime.formats.relative.php) (e.g. `1 day`, `last week`, `-2 weeks`), it always uses the current system time as the reference point.

While in many cases this isn't an issue, I've been using relative date formats in my YAML fixtures for unit tests. In one recent case, I was attempting to filter a list down to fixtures that had a date field set in the past, but the filter was pulling in more `DataObjects` than intended because at least 10 seconds had passed between me calling `SS_Datetime::set_mock_now(SS_Datetime::now())` in the test's `setUp()` method and the text fixtures actually being created.

I'll work around this by setting fixed dates and times in my YAML fixtures, as I imagine most people do anyway and I'm just an edge case.

## Steps to Reproduce

Create a unit test case:
* `setUpOnce` calls `SS_Datetime::set_mock_now()`
* `tearDownOnce` calls `SS_Datetime::clear_mock_now()`
* Create a single unit test which does the following:
  * `sleep` for five seconds
  * Create an expected timestamp string by creating a new `SS_Datetime` instance, setting its value to one second from the mocked now time, and calling `getValue()` on it.
  * Create an `SS_Datetime` field with its value set to one second from the mocked now time.
  * Create an `SS_Datetime` field with its value set to one second from now.
  * Assert that the first field's value equals the expected timestamp.
  * Assert that the second field's value equals the expected timestamp.

This unit test _should_ pass, but it will fail.

```php
<?php

class RelativeTimeTest extends SapphireTest {
	public function setUpOnce() {
		SS_Datetime::set_mock_now(SS_Datetime::now());
		parent::setUpOnce();
	}
	public function tearDownOnce() {
		parent::tearDownOnce();
		SS_Datetime::clear_mock_now();
	}
	public function testRelativeTime() {
		sleep(5);

		$mockedNowPlusOne = SS_Datetime::now()->getValue() . ' +1 sec';
		$nowPlusOne = '1 sec';

		$fieldWithMockedNow = SS_Datetime::create();
		$fieldWithMockedNow->setValue($mockedNowPlusOne);
		$expected = $fieldWithMockedNow->getValue();

		$withReference = SS_Datetime::create();
		$withReference->setValue($mockedNowPlusOne);
		$withoutReference = SS_Datetime::create();
		$withoutReference->setValue($nowPlusOne);

		//This assertion will pass
		$this->assertEquals($expected, $withReference->getValue());
		//This assertion will fail, and it should not.
		$this->assertEquals($expected, $withoutReference->getValue());
	}
}

**Comment by maxime-rainville** [comment]
Hi @p0lar-bear,

I just ran the unit test you provided with a clean Silverstripe 3.6 install on PHP7.1 and both assertions passed for me.

Can you have second look and let me know if there's any other condition required to trigger the issue?

**Comment by dhensby** [comment]
should `$nowPlusOne = '1 sec';` be `$nowPlusOne = '+1 sec';` ?

**Comment by p0lar-bear** [comment]
@dhensby In that case it doesn't matter, as PHP interprets that the same way as `'now +1 sec'`. Whereas `$mockedNowPlusOne` ends up being something along the lines of `'2018-04-17 15:47:30 +1 sec'`

@maxime-rainville Are you able to reproduce results with a fresh Silverstripe 3.5? I will admit that 3.6 is an assumption because `Datetime.php` doesn't change at all between those versions. It may be worth mentioning that I am running on PHP 5.6 and I don't have a means to run this on 7.1?

**Comment by tractorcow** [comment]
The problem is this code that parses string values.

```
if(is_string($value)) {
    try {
        $date = new DateTime($value);
```

There is no way to inject `now' (mocked or otherwise) when passing a relative date time string to the DateTime constructor.

Possible fix is to substitute this with strtotime().

Also your test below is not good:

```
$this->assertEquals($expected, $withReference->getValue());
```

Both your $expected and $withReference are constructed exactly the same way, so those would always be the same. :)

I would recommend to just not use relative time as a datetime constructor argument. The fact that this works is more of a coincidence than a design on the part of the framework.

Framework 4 doesn't let you do this in any case, since it validates the source string and errors on ambiguous dates. It will error with `"Invalid date: '$value'. Use " . self::ISO_DATE . " to prevent this error."` This seems to me like it is unlikely we'll make any changes to this in 3.x also, I'm afraid.

Long story short, don't rely on relative dates if you can help it, but if you do you'll need to handle the manipulation in user code.

**Comment by p0lar-bear** [comment]
Noted, I'll convert my fixture timestamps to fixed times, as it's a better practice.

---

Though, I do want to share something I was going to put into a PR that seemed to fix my broken test, and seemed to make logical sense. I don't present this as if to suggest that the decision to nix relative format support in 4.x was wrong, as I wouldn't be surprised if this approach has an oversight that would cause issues I can't forsee, but this could be useful for other things, maybe.

> There is no way to inject `now' (mocked or otherwise) when passing a relative date time string to the DateTime constructor.

On the contrary - as stated in [PHP: Relative Formats](http://php.net/manual/en/datetime.formats.relative.php), in a date string with a relative format, PHP will always process non-relative statements _before_ relative ones. I got the idea to address that shortcoming by taking the relative format and appending it to the value of `SS_Datetime::now()`. The only problem there being, how does one detect if a date string is relative?

[`date_parse()`](http://php.net/manual/en/function.date-parse.php) seems to be the best way to go about doing it. It outputs an associative array of the parsed timestamp's different parts, as well as warnings/errors. In addition, _if the parsed timestamp has a relative format, it adds a `relative` key with information about the relative time data.`

If there's a fixed date/time in the parsed timestamp, that fixed time is parsed independently of the relative time. So, `2018-04-18 12:47:23 1 hour` will get this output from `date_parse()`:
```
Array
(
    [year] => 2018
    [month] => 4
    [day] => 18
    [hour] => 12
    [minute] => 47
    [second] => 23
    [fraction] => 0
    [warning_count] => 0
    [warnings] => Array ()
    [error_count] => 0
    [errors] => Array ()
    [is_localtime] =>
    [relative] => Array (
            [year] => 0
            [month] => 0
            [day] => 0
            [hour] => 1
            [minute] => 0
            [second] => 0
        )
)
```
If the fixed time is `now` (i.e. either `'now 1 week'` or `'1 week`), the `year`, `month`, `day`, `hour`, `minute`, `second`, and `fraction` values will all be set to `false`. So, right before it attempts to parse the value with the `DateTime` constructor, I put this in:
```php
//...
} elseif(is_string($value)) {
	// $this->value = date('Y-m-d H:i:s', strtotime($value));
	// Check incoming date string to see if it's relative to "now"
	$parsedDate = date_parse($value);
	if (array_key_exists('relative', $parsedDate) &&
		!$parsedDate['year'] && !$parsedDate['month'] && !$parsedDate['day'] && !$parsedDate['hour'] &&
		!$parsedDate['minute'] && !$parsedDate['second'] && !$parsedDate['fraction']
	) {
		//Giving date_parse() no date or 'now' results in all of the keys listed above being false.
		$value = self::now() . ' ' . $value;
	}
	try{
		$date = new DateTime($value);
//...
```
